pkg/binlog/binlog.go

```diff
@@ -46,11 +47,12 @@ func (c *Client) hookAddr(addr string) string {
 	return addr
 }
 
-// NewBinlogClient create a Client.
+// NewBinlogClient create a Client and return an error if the underlying conn is not up within 5 seconds.
 func NewBinlogClient(pdEndpoint []string, tlsConfig *tls.Config) (*Client, error) {
 	etcdClient, err := clientv3.New(clientv3.Config{
 		Endpoints:   pdEndpoint,
 		DialTimeout: time.Second * 5,
+		DialOptions: []grpc.DialOption{grpc.WithBlock()},
```

Comment:
Even if we succeed here, we still have to add timeout for the operations, e.g. https://github.com/pingcap/tidb-operator/blob/master/pkg/binlog/binlog.go#L181.
You can refer to the logic here https://github.com/pingcap/tidb-operator/blob/master/pkg/pdapi/pdetcd.go#L69.

Reply:
BTW, please also add timout to the HTTP client in case someone else use it later.

Reply:
The `DialTimeout` should be OK here, right? So no need to add the `DialOptions` here, please check the example here https://pkg.go.dev/go.etcd.io/etcd/clientv3#pkg-overview and the description for `DialTimeout`.

Reply:
As mentioned in this pr's comment, even though `DialTimeout` has been set, `clientv3.New()` returns immediately.
This seems like the behavior changed after v3.3  https://github.com/etcd-io/etcd/issues/9877#issuecomment-399262512.
Also checkout the comments for field `DialOptions` here. https://pkg.go.dev/go.etcd.io/etcd@v3.3.27+incompatible/clientv3#Config